[Form Control Refresh] Number fields do not have the same width as text fields

lilyspiniolas opened the pull request:
#### 476f1ef1dbfc800d96c21282b529e095ed49a15e
<pre>
[Form Control Refresh] Number fields do not have the same width as text fields
<a href="https://bugs.webkit.org/show_bug.cgi?id=296340">https://bugs.webkit.org/show_bug.cgi?id=296340</a>
<a href="https://rdar.apple.com/156408135">rdar://156408135</a>

Reviewed by Aditya Keerthi.

With form control refresh enabled, number fields previously did not have the
same width as text fields due a difference in inline-end padding. These changes
resolve this issue by adding the difference to the maximum logical width of
numberfields during layout.

Added layout tests for inputs of type `email`, `number`, `password`, `tel`,
and `url` to make sure each input type has the same logical width as text
inputs for all writing-modes and directions. These additional tests also
ensure that the listed input types have the same logical width as text inputs
when both have equal inline-padding.

fast/forms/input-password-logical-widths.html does not pass on iOS due to
an issue where password inputs have a larger logical width than other
text-related inputs for the RTL state.

* LayoutTests/fast/forms/input-email-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-email-logical-widths.html: Added.
* LayoutTests/fast/forms/input-number-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-number-logical-widths.html: Added.
* LayoutTests/fast/forms/input-password-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-password-logical-widths.html: Added.
* LayoutTests/fast/forms/input-tel-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-tel-logical-widths.html: Added.
* LayoutTests/fast/forms/input-url-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-url-logical-widths.html: Added.
* LayoutTests/fast/forms/input-widths.html:
* LayoutTests/fast/forms/resources/input-logical-widths-helper.js: Added.
(assertLogicalWidthsEqual):
(setWritingMode):
(testVerticalRL):
(testSidewaysLR):
(testSidewaysRL):
(test):
(paddingMsg.string_appeared_here.inputs.forEach):
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::adjustedMaximumLogicalWidthForControl const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::paddingBoxForNumberField):
(WebCore::applyEmPaddingForNumberField):
(WebCore::RenderThemeCocoa::adjustTextFieldStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustedMaximumLogicalWidthForControl const):

Canonical link: <a href="https://commits.webkit.org/297890@main">https://commits.webkit.org/297890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcb106250f06f60782c8e6bb60060c048540ae55

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113283 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119491 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63967 "Built successfully") | ✅ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33671 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41581 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86217 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ✅ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116230 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101904 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66544 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26162 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20031 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63245 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20106 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122708 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95065 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94811 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24192 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39958 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17776 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36497 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40247 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45746 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39888 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43221 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41625 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/input-logical-widths.html:
```diff
@@ -0,0 +1,131 @@
+var baseWidth = document.getElementById('text').offsetWidth;
+function assertLogicalWidthsEqual(isVertical) {
+	if (isVertical) {
+		shouldBe('document.getElementById("email").offsetHeight', 'baseWidth');
```
Use spaces instead of tabs (throughout this file).

pxlcoder commented on LayoutTests/fast/forms/input-logical-widths.html:
```diff
@@ -0,0 +1,131 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../resources/js-test-pre.js"></script>
```
No need for pre/post, just use `js-test.js`.

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -3744,6 +3749,32 @@ static inline bool shouldAdjustTextControlInnerElementStyles(const RenderStyle&
     return { 0, 0 };
 }
+
+float RenderThemeCocoa::adjustedMaximumLogicalWidthForControl(const RenderStyle& style, const Element& element, float maximumLogicalWidth) const
```
Should this method check that native appearance is used?

lilyspiniolas replied:
Yes, added a check

pxlcoder commented on LayoutTests/fast/forms/input-widths.html:
```diff
@@ -1,35 +1,36 @@
-<p id="description">Text field types should have the same default widths.</p>
-<div id="console"></div>
+ <html>
+ <head>
```
What's up with these single spaces at the beginning of each line?

pxlcoder commented on LayoutTests/fast/forms/input-widths.html:
```diff
@@ -1,35 +1,36 @@
-<div id="console"></div>
+ <html>
+ <head>
+ <script src="../../resources/js-test-pre.js"></script>
```
Same comment about using `js-test.js`.

pxlcoder commented on LayoutTests/fast/forms/input-widths.html:
```diff
@@ -1,35 +1,36 @@
+ <script src="../../resources/js-test-post.js"></script>
+ </body>
+ </html>
+ 
```
Newline weirdness down here.

pxlcoder commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -8219,6 +8219,9 @@ fast/forms/ios/form-control-refresh/color/paint-within-box.html [ ImageOnlyFailu
 imported/w3c/web-platform-tests/css/css-view-transitions/fractional-translation-from-position.html [ ImageOnlyFailure ]
 
+# <rdar://156470462> RTL inputs with type=password have larger logical width than RTL inputs with type number, url, tel, & text
+fast/forms/input-logical-widths.html [ Failure ]
```
It's not great that we need to mark the entire test as a failure because of one input type? Can the test be split up?

lilyspiniolas replied:
Done!

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -3744,6 +3749,35 @@ static inline bool shouldAdjustTextControlInnerElementStyles(const RenderStyle&
+        const auto paddingBox = paddingBoxForNumberField(style, &element);
+
+        const auto paddingEdgeInlineStart = isHorizontal ? (isInlineFlipped ? paddingBox.right() : paddingBox.left()) : (isInlineFlipped ? paddingBox.bottom() : paddingBox.top());
+        const auto paddingEdgeInlineEnd = isHorizontal ? (isInlineFlipped ? paddingBox.left() : paddingBox.right()) : (isInlineFlipped ? paddingBox.top() : paddingBox.bottom());
```
You should be able to use the logical variants, `start` and `end` here. See `RectEdges.h`.

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -3742,6 +3747,35 @@ static inline bool shouldAdjustTextControlInnerElementStyles(const RenderStyle&
+    const auto isHorizontal = writingMode.isHorizontal();
+    const auto isInlineFlipped = writingMode.isInlineFlipped();
+
+    const auto inlineEndPaddingExplicitlySet = isHorizontal ? (isInlineFlipped ? style.hasExplicitlySetPaddingLeft() : style.hasExplicitlySetPaddingRight()) : (isInlineFlipped ? style.hasExplicitlySetPaddingTop() : style.hasExplicitlySetPaddingBottom());
```
This would be more readable as an immediately invoked lambda.

lilyspiniolas committed:
[Form Control Refresh] Number fields do not have the same width as text fields
https://bugs.webkit.org/show_bug.cgi?id=296340
rdar://156408135

Reviewed by Aditya Keerthi.

With form control refresh enabled, number fields previously did not have the
same width as text fields due a difference in inline-end padding. These changes
resolve this issue by adding the difference to the maximum logical width of
numberfields during layout.

Added layout tests for inputs of type `email`, `number`, `password`, `tel`,
and `url` to make sure each input type has the same logical width as text
inputs for all writing-modes and directions. These additional tests also
ensure that the listed input types have the same logical width as text inputs
when both have equal inline-padding.

fast/forms/input-password-logical-widths.html does not pass on iOS due to
an issue where password inputs have a larger logical width than other
text-related inputs for the RTL state.

* LayoutTests/fast/forms/input-email-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-email-logical-widths.html: Added.
* LayoutTests/fast/forms/input-number-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-number-logical-widths.html: Added.
* LayoutTests/fast/forms/input-password-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-password-logical-widths.html: Added.
* LayoutTests/fast/forms/input-tel-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-tel-logical-widths.html: Added.
* LayoutTests/fast/forms/input-url-logical-widths-expected.txt: Added.
* LayoutTests/fast/forms/input-url-logical-widths.html: Added.
* LayoutTests/fast/forms/input-widths.html:
* LayoutTests/fast/forms/resources/input-logical-widths-helper.js: Added.
(assertLogicalWidthsEqual):
(setWritingMode):
(testVerticalRL):
(testSidewaysLR):
(testSidewaysRL):
(test):
(paddingMsg.string_appeared_here.inputs.forEach):
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::adjustedMaximumLogicalWidthForControl const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::paddingBoxForNumberField):
(WebCore::applyEmPaddingForNumberField):
(WebCore::RenderThemeCocoa::adjustTextFieldStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustedMaximumLogicalWidthForControl const):

Canonical link: https://commits.webkit.org/297890@main
